docs/features/advanced-usage/prompts.mdx

```diff
@@ -0,0 +1,468 @@
+    }
+  ]
+}
+```
+</CodeGroup>
+
+### Message Handling
+
+Messages work differently than other parameters. Instead of overriding, runtime messages are **appended** to the saved prompt messages. This allows you to:
+
+- Define consistent system prompts and example conversations in your saved prompt
+- Add dynamic user messages at runtime
+- Build multi-turn conversations that maintain context
+
+<Warning>
+Runtime messages are always appended to the end of your saved prompt messages. Make sure your saved prompt structure accounts for this behavior.
+</Warning>
+
+### Override Examples
+
+<Tabs>
+<Tab title="Temperature Override">
+```typescript
+// Saved prompt has temperature: 0.8
+const response = await openai.chat.completions.create({
+  prompt_id: "abc123",
+  temperature: 0.2, // Uses 0.2, not 0.8
+  inputs: { topic: "AI safety" }
+});
+```
+</Tab>
+
+<Tab title="Max Tokens Override">
+```typescript
+// Saved prompt has max_tokens: 500
+const response = await openai.chat.completions.create({
+  prompt_id: "abc123", 
+  max_tokens: 1500, // Uses 1500, not 500
+  inputs: { complexity: "detailed" }
+});
+```
+</Tab>
+
+<Tab title="Response Format Override">
+```typescript
+// Saved prompt has no response format
+const response = await openai.chat.completions.create({
+  prompt_id: "abc123",
+  response_format: { type: "json_object" }, // Adds JSON formatting
+  inputs: { data_type: "user_preferences" }
+});
+```
+</Tab>
+</Tabs>
+
+<Note>
+This compilation approach gives you the flexibility to have consistent prompt templates while still allowing runtime customization for specific use cases.
+</Note>
+
+## Variables
+
+Variables make your prompts dynamic and reusable. Define them once in your prompt template, then provide different values at runtime without changing your code.
+
+### Variable Syntax
+
+Variables use the format `{{hc:name:type}}` where:
+- `name` is your variable identifier
+- `type` defines the expected data type
+
+<CodeGroup>
+```text Basic Examples
+{{hc:customer_name:string}}
+{{hc:age:number}}
+{{hc:is_premium:boolean}}
+{{hc:context:any}}
+```
+
+```text In Prompt Templates
+You are a helpful assistant for {{hc:company:string}}.
+
+The customer {{hc:customer_name:string}} is {{hc:age:number}} years old.
+Premium status: {{hc:is_premium:boolean}}
+
+Additional context: {{hc:context:any}}
+```
+</CodeGroup>
+
+### Supported Types
+
+| Type | Description | Example Values | Validation |
+|------|-------------|----------------|------------|
+| `string` | Text values | `"John Doe"`, `"Hello world"` | None |
+| `number` | Numeric values | `25`, `3.14`, `-10` | AI Gateway type-checking |
+| `boolean` | True/false values | `true`, `false`, `"yes"`, `"no"` | AI Gateway type-checking |
+| `your_type_name` | Any data type | Objects, arrays, strings | None |
+
+<Warning>
+Only `number` and `boolean` types are validated by the AI Gateway, which will accept strings for any input as long as they can be converted to valid values.
```

Comment:
```suggestion
Only `number` and `boolean` types are validated by the Helicone AI Gateway, which will accept strings for any input as long as they can be converted to valid values.
```

Let's try to always call it Helicone AI Gateway so it's clear what we're talking about